### ISSUE TITLE
Luquillo/Miguel: Implementing SamplingFeature geometry types beyond points, in PostGIS

### ISSUE THREAD

**Issue by emiliom** [ODM2/ODM2#111]
I'm following up on a comment @miguelcleon made at the Luquillo CZO 1:1 consultation this week. Miguel is already actively using ODM2 with PostgreSQL and PostGIS, but currently is handling only simple-Point [SamplingFeature](http://odm2.github.io/ODM2/schemas/ODM2_Current/diagrams/ODM2SamplingFeatures.html) geometry types. I said I'd be happy to help explore the implementation of more complex types (mainly lines/vectors, but possibly polygons and multi-points?) in PostGIS.

I'm doing this as a github issue so we keep a public record for easy future reference.

So, Miguel, do you have specific applications in mind already? I'm guessing that one of them might be for transects or trajectories [result type?](http://vocabulary.odm2.org/resulttype/)

**Comment by miguelcleon** [comment]
Yes, we certainly have transect result type data which would be great to be able to describe. I'm also curious about storing shape file data, a simple example might be a shape file describing categorical forest types for a landscape.

**Comment by emiliom** [comment]
Miguel, I'll follow up tomorrow or Friday. For now, I'm just pasting some references that I'll refer back to later, so I don't forget. FYI, I'll probably refer to these when we talk about storing shape file data like categorical forest types.

PostgreSQL/PostGIS OGR Foreign Data Wrapper (FDW):
- http://boundlessgeo.com/2014/12/ogr-fdw-ftw/
- http://www.postgresonline.com/journal/archives/339-OGR-foreign-data-wrapper-on-Windows-first-taste.html

**Comment by emiliom** [comment]
**I'll start with your second question:**
> I'm also curious about storing shape file data, a simple example might be a shape file describing categorical forest types for a landscape. This would be nice to eventually map, with other point or transect data overlain onto it. 

It sounds like you're asking about general storage of relevant spatial layers, for convenience and ability to link with the observed data. If you already use PostGIS a lot (or want to), there can definitely be advantages to this. But it's important to keep in mind that ODM2 doesn't directly address data related to environmental features characterizing the sampling feature, unless those environmental features are themselves your observed properties. For example, if the observations involve soil properties, a vegetation type is a very helpful and common characteristic to store, but a land cover map of the area is not what's being observed; it's valuable ancillary information or context.

I'm assuming that's the general application you have in mind. In that case, ODM2 doesn't have anything specific to handle this. Storing a geospatial polygon layer of land cover types in PostGIS in the same ODM2 PostgreSQL database would be just a convenience for data management and access, and possibly for geospatial operations if you're interested in that. Nothing more. But if that's what you want to do, I recommend you store those ancillary geospatial layers in a separate schema that you create for that purpose, keeping the layers separate from the ODM2 objects (tables, views, etc). You can import shape files directly into PostGIS in that separate schema, as basically copies of your shape files. Lots of software can do that (PostGIS tools, QGIS, ogr2ogr, etc). Alternatively, if you plan to keep using the shape files directly, you might consider setting up an OGR Foreign Data Wrapper (FDW) to define a linkage directly to the shape files (almost like a view) w/o actually importing a copy; see the two links I posted in my previous comment.

FYI, if those environmental characteristics are your observed data, you can use the Taxonomic Classifiers in ODM2. In that case you will no longer have a simple geospatial layer, but rather a highly restructured version of it that may involve considerable effort to put back together as, say, a simple polygon GIS layer. You may not even be dealing with polygons anymore, if that's not how the characteristic was observed.

**I'll comment soon on your first question,** the one that involves ODM2 directly: representing a transect SamplingFeature.

**Comment by emiliom** [comment]
**Getting to your first question, how to represent transect data in ODM2.**

There are two concepts of "transect" in ODM2:
- [A transectCoverage Result](http://vocabulary.odm2.org/resulttype/transectCoverage/) type.
- [A transect (line vector geometry) SamplingFeature](http://vocabulary.odm2.org/samplingfeaturetype/transect/) type.

The two can be used together; if you use a `transectCoverage` Result type, I think you would have to define a `transect` SamplingFeature as well, otherwise your representations would be inconsistent.

But an alternative to consider is representing your observed results (the soil measurements carried out at individual locations along the transect) using simple `measurement` Result type, with each location associated with a corresponding Point SamplingFeature (which could be defined as a `site` SamplingFeature, if it makes sense in your context). The concept and linear geometry of the transect would then be stored as a separate `transect` SamplingFeature built up as a line that connects all the point SamplingFeature records in the transect, and the linkage to those points would be via [`RelatedFeatures` entries that establishes parent (transect) - children (points) relationships](http://odm2.github.io/ODM2/schemas/ODM2_Current/diagrams/ODM2SamplingFeatures.html).

For a soils transect, I would recommend this second approach, for a couple of reasons:
- You probably are very interested in the identity of the individual sampling points; I suspect you even have location codes or names for them, and you probably will want to query on those locations, show them in maps (together with the transect line), etc. Handling your transects as `transectCoverage` Result type breaks the direct and unambiguous link between an individual result and an individual point SamplingFeature definition. You would have to query the `TransectResultValues` table, where your result values are stored, to extract, say, all data at a location or query on a location code. This could be slow, and it's more liable to inconsistent data entry b/c there isn't a single record storing the location and location code.
- By reusing a result values table you're already using (`MeasurementResultValues`), your data management and Admin UI application code may be a bit less complicated than if you use `TransectResultValues`.

The use of `transectCoverage` result type is best suited for observations done by a platform that is repeating exactly the same kind of observations, with the same instruments, in an identical way, and at a fairly high spatial sampling frequency. For example, "continuous" sampling from a boat with a sensor hanging on the side. @horsburgh showed us today a cool example of this from an instrument and floating platform his group has put together.

I can share with you some PostGIS SQL code to create a line SamplingFeature from point SamplingFeatures. Bear in mind that ideally you would also need to create an `Action` that corresponds to the transect SamplingFeature, which would be a parent action to the actions related to sampling at each location.
